File: clients/go/consumer.go
// pgque-go -- Go client for PgQue
// Copyright 2026 Nikolay Samokhvalov. Apache-2.0 license.

package pgque

import (
	"context"
	"fmt"
	"log"
	"time"
)

// HandlerFunc processes a single message. Returning a non-nil error
// causes the Consumer to issue a per-message Nack for that message
// (routing it to retry_queue or, once max_retries is exceeded, to the
// dead_letter table). Other messages in the same batch are still
// dispatched to their handlers; the batch as a whole is acked only if
// every required Nack succeeded.
type HandlerFunc func(ctx context.Context, msg Message) error

// consumerBackend is the subset of Client used by Consumer. Defining
// it as an interface keeps the Consumer testable: a stub backend can
// simulate Nack failures without a live database.
type consumerBackend interface {
	Receive(ctx context.Context, queue, consumer string, maxMessages int) ([]Message, error)
	Ack(ctx context.Context, batchID int64) error
	Nack(ctx context.Context, batchID int64, msg Message, opts ...NackOption) error
}

// Consumer polls a queue and dispatches messages to registered
// handlers. Create one via Client.NewConsumer.
type Consumer struct {
	backend       consumerBackend
	queue         string
	name          string
	pollInterval  time.Duration
	maxMessages   int
	handlers      map[string]HandlerFunc
	unknownPolicy UnknownHandlerPolicy
}

// Handle registers fn as the handler for messages whose Type matches
// eventType. Messages with no registered handler are dispatched per
// the consumer's UnknownHandlerPolicy: by default each is Nack'd
// individually (routing it to retry_queue or eventually the DLQ);
// with WithUnknownHandlerPolicy(AckUnknown) they are logged and
// silently skipped instead.
func (c *Consumer) Handle(eventType string, fn HandlerFunc) {
	c.handlers[eventType] = fn
}

// dispatchWithRecover calls fn and converts any panic into a non-nil
// error so that the caller can nack the message and keep polling.
func (c *Consumer) dispatchWithRecover(ctx context.Context, fn HandlerFunc, msg Message) (retErr error) {
	defer func() {
		if r := recover(); r != nil {
			retErr = fmt.Errorf("handler panic: %v", r)
		}
	}()
	return fn(ctx, msg)
}

// Start begins the poll loop and blocks until ctx is cancelled. On
// receive errors it logs and retries after the configured poll
// interval.
//
// Per-batch dispatch semantics:
//
//   - Each message is delivered to its registered handler. If the
//     handler returns a non-nil error or panics, the message is
//     individually Nack'd (routed to retry_queue, eventually the DLQ).
//   - Messages with no registered handler are Nack'd or skipped
//     depending on the configured UnknownHandlerPolicy (default: Nack).
//   - If every required Nack call succeeded (or none were needed), the
//     batch is Ack'd. If any Nack fails, the batch is left unacked so
//     that PgQue redelivers it on the next Receive — losing the Nack
//     would otherwise mean losing the failure information for that
//     message.
func (c *Consumer) Start(ctx context.Context) error {
	for {
		select {
		case <-ctx.Done():
			return ctx.Err()
		default:
		}

		msgs, err := c.backend.Receive(ctx, c.queue, c.name, c.maxMessages)
		if err != nil {
			log.Printf("pgque: receive error: %v", err)
			select {
			case <-ctx.Done():
				return ctx.Err()
			case <-time.After(c.pollInterval):
			}
			continue
		}

		if len(msgs) == 0 {
			select {
			case <-ctx.Done():
				return ctx.Err()
			case <-time.After(c.pollInterval):
			}
			continue
		}

		var batchID int64
		nackFailed := false
		for _, msg := range msgs {
			batchID = msg.BatchID
			handler, ok := c.handlers[msg.Type]
			if !ok {
				if c.unknownPolicy == AckUnknown {
					log.Printf("pgque: no handler registered for event type %q, skipping message %d (AckUnknown policy)", msg.Type, msg.MsgID)
					continue
				}
				log.Printf("pgque: no handler registered for event type %q, nacking message %d", msg.Type, msg.MsgID)
				if nackErr := c.backend.Nack(ctx, batchID, msg); nackErr != nil {
					log.Printf("pgque: nack error for unhandled type %s: %v", msg.Type, nackErr)
					nackFailed = true
				}
				continue
			}
			if handlerErr := c.dispatchWithRecover(ctx, handler, msg); handlerErr != nil {
				log.Printf("pgque: handler error for %s: %v", msg.Type, handlerErr)
				if nackErr := c.backend.Nack(ctx, batchID, msg); nackErr != nil {
					log.Printf("pgque: nack error for %s: %v", msg.Type, nackErr)
					nackFailed = true
				}
				continue
			}
		}

		if batchID != 0 {
			if nackFailed {
				log.Printf("pgque: skipping ack for batch %d due to prior nack failures; PgQue will redeliver", batchID)
				continue
			}
			if err := c.backend.Ack(ctx, batchID); err != nil {
				log.Printf("pgque: ack error: %v", err)
			}
		}
	}
}

File: clients/typescript/src/consumer.ts
// pgque -- TypeScript client for PgQue
// Copyright 2026 Nikolay Samokhvalov. Apache-2.0 license.

import type { Client } from './client.js';
import type { ConsumerOptions, HandlerFunc, Message } from './types.js';

/**
 * Default `maxMessages` for the high-level Consumer. PostgreSQL `int4` max
 * (`2^31 - 1`); request the whole PgQ batch by default so a subsequent
 * `pgque.ack(batch_id)` does not strand events the client never saw.
 */
export const DEFAULT_MAX_MESSAGES = 2_147_483_647;

/**
 * High-level consumer that polls `pgque.receive`, dispatches each message
 * to a per-event-type handler, and finalizes the batch with `ack` (or
 * per-message `nack` on handler failure / unknown event type).
 *
 * Usage:
 * ```ts
 * const consumer = client.newConsumer('orders', 'order_worker');
 * consumer.handle('order.created', async (msg) => { ... });
 *
 * const ac = new AbortController();
 * await consumer.start(ac.signal);
 * ```
 */
export class Consumer {
  private readonly handlers = new Map<string, HandlerFunc>();
  private readonly pollIntervalMs: number;
  private readonly maxMessages: number;
  private readonly retryAfter: number;
  private readonly unknownHandlerPolicy: 'ack' | 'nack';
  private readonly logger: Pick<Console, 'warn' | 'error'>;
  private readonly subconsumer: string | undefined;
  private readonly deadInterval: string | undefined;

  /** @internal — use {@link Client.newConsumer}. */
  constructor(
    private readonly client: Client,
    private readonly queue: string,
    private readonly name: string,
    opts: ConsumerOptions = {},
  ) {
    this.pollIntervalMs = opts.pollInterval ?? 30_000;
    this.maxMessages = opts.maxMessages ?? DEFAULT_MAX_MESSAGES;
    this.retryAfter = opts.retryAfter ?? 60;
    if (!Number.isFinite(this.retryAfter) || this.retryAfter < 0) {
      throw new Error('pgque: retryAfter must be a non-negative finite number of seconds');
    }
    this.unknownHandlerPolicy = opts.unknownHandlerPolicy ?? 'nack';
    this.logger = opts.logger ?? console;
    this.subconsumer = opts.subconsumer;
    this.deadInterval = opts.deadInterval;
    if (this.deadInterval !== undefined && this.subconsumer === undefined) {
      throw new Error(
        'pgque: deadInterval requires subconsumer; pass { subconsumer, deadInterval } together',
      );
    }
  }

  /** Register a handler for `eventType`. Replaces any previous handler. */
  handle(eventType: string, fn: HandlerFunc): void {
    this.handlers.set(eventType, fn);
  }

  /**
   * Start the poll loop. Resolves when `signal` is aborted; rejects only
   * on terminal errors that should bubble up (the routine `Receive`/`Ack`
   * errors are logged and the loop continues).
   *
   * **Abort granularity:** aborting the signal interrupts the inter-poll
   * `sleep()` immediately, but does **not** cancel an in-flight
   * `client.receive()` call. If a `receive()` round-trip is in progress
   * when the signal fires, the loop will drain that call to completion
   * before exiting.
   */
  async start(signal?: AbortSignal): Promise<void> {
    while (!signal?.aborted) {
      let msgs: Message[];
      try {
        msgs =
          this.subconsumer !== undefined
            ? await this.client.receiveCoop(this.queue, this.name, this.subconsumer, {
                maxMessages: this.maxMessages,
                ...(this.deadInterval !== undefined ? { deadInterval: this.deadInterval } : {}),
              })
            : await this.client.receive(this.queue, this.name, this.maxMessages);
      } catch (err) {
        this.logger.error(`pgque: receive error: ${formatErr(err)}`);
        await sleep(this.pollIntervalMs, signal);
        continue;
      }

      if (msgs.length === 0) {
        await sleep(this.pollIntervalMs, signal);
        continue;
      }

      let batchId: bigint | null = null;
      let anyNackFailed = false;
      for (const msg of msgs) {
        batchId = msg.batchId;
        const handler = this.handlers.get(msg.type);
        if (!handler) {
          if (this.unknownHandlerPolicy === 'ack') {
            this.logger.warn(
              `pgque: no handler registered for event type "${msg.type}", acking msg ${msg.msgId} (unknownHandlerPolicy='ack')`,
            );
            // Fall through; the batch ack at the end of the loop covers it.
            continue;
          }
          this.logger.warn(
            `pgque: no handler registered for event type "${msg.type}", nacking msg ${msg.msgId}`,
          );
          if (!(await this.tryNack(batchId, msg, `unknown event type: ${msg.type}`))) {
            anyNackFailed = true;
          }
          continue;
        }
        try {
          await handler(msg);
        } catch (err) {
          this.logger.error(`pgque: handler error for "${msg.type}": ${formatErr(err)}`);
          if (!(await this.tryNack(batchId, msg, `handler error: ${formatErr(err)}`))) {
            anyNackFailed = true;
          }
        }
      }

      if (batchId !== null) {
        if (anyNackFailed) {
          // At least one required nack failed. Skip ack so PgQ redelivers
          // the batch instead of advancing the consumer past messages we
          // couldn't route. The batch is unfinished, so `next_batch` would
          // return it again immediately — sleep one poll interval before
          // re-polling to avoid a hot loop that re-runs every handler.
          this.logger.error(
            `pgque: skipping ack for batch ${batchId}; one or more nacks failed and the batch will be redelivered`,
          );
          await sleep(this.pollIntervalMs, signal);
        } else {
          try {
            const n = await this.client.ack(batchId);
            if (n === 0) {
              this.logger.warn(
                `pgque: ack batch ${batchId} returned 0 — stale or double ack (batch already finished or not found)`,
              );
            }
          } catch (err) {
            // Unfinished batch: same redelivery situation as the failed-nack
            // path above, so back off one poll interval before re-polling.
            this.logger.error(`pgque: ack error: ${formatErr(err)}`);
            await sleep(this.pollIntervalMs, signal);
          }
        }
      }
    }
  }

  /** Returns true if the nack succeeded, false if it threw (and was logged). */
  private async tryNack(batchId: bigint, msg: Message, reason: string): Promise<boolean> {
    try {
      await this.client.nack(batchId, msg, { retryAfter: this.retryAfter, reason });
      return true;
    } catch (err) {
      this.logger.error(`pgque: nack error for "${msg.type}": ${formatErr(err)}`);
      return false;
    }
  }
}

function formatErr(err: unknown): string {
  if (err instanceof Error) return err.message;
  return String(err);
}

function sleep(ms: number, signal?: AbortSignal): Promise<void> {
  return new Promise((resolve) => {
    if (signal?.aborted) {
      resolve();
      return;
    }
    const timer = setTimeout(() => {
      signal?.removeEventListener('abort', onAbort);
      resolve();
    }, ms);
    const onAbort = (): void => {
      clearTimeout(timer);
      resolve();
    };
    signal?.addEventListener('abort', onAbort, { once: true });
  });
}
